cleanup redundant text attributes when delete attributes

dkuhnert opened the pull request:
Fixes #392

I was able to track it down to some redundant formatting that was left from the format function.

I expected that the text contains following contents after Change 2 (`text.format(16, 4, { bold: null })`) from the issue:

```js
ContentString { str: 'Attack ships ' }
ContentFormat { key: 'bold', value: true }
ContentString { str: 'on ' }
ContentFormat { key: 'bold', value: null }
ContentString { str: 'fire' }
ContentString { str: ' off the shoulder of Orion.' }
```

But instead got this contents:

```js
ContentString { str: 'Attack ships ' }
ContentFormat { key: 'bold', value: true }
ContentString { str: 'on ' }
ContentFormat { key: 'bold', value: null }
ContentString { str: 'fire' }
ContentFormat { key: 'bold', value: true } // redundant
ContentFormat { key: 'bold', value: null } // redundant
ContentString { str: ' off the shoulder of Orion.' }
```

Which contains redundant formatting contents.
I added an detection for this which fixes the issue.

dmonad commented on src/types/YText.js:
```diff
@@ -164,6 +164,20 @@ const insertNegatedAttributes = (transaction, parent, currPos, negatedAttributes
+    // check if we really need to create attributes
+    //  (the attribute may be set the desired value already)
+    let n = currPos.right
+    while(
```
I think the problem is that we want to insert a negated attribute, but the negated attribute might already exist in one of the following `ContentFormat` items. So, when we insert the same `ContentFormat` twice, we end up with the undesired result.

The ideal solution, it seems to me is to avoid inserting the negated attribute if it is already inserted in one of the following `ContentFormat` items.

This approach, however, seems to iterate through the whole list of `ContentFormat` for each entry in `negatedAttributes` (O(N•M)). Then it deletes the entries that we want to insert anyway.

I suggest that you simply `return` when you find an item that was already inserted with the same key (the values shouldn't matter here). There is no need to delete it and `return` (we would end up without negatedAttributes if we deleted an item, but don't insert anything).

dkuhnert replied:
This makes sense, i changed the pull request to use the previous cleanup loop, this will now also check the attributes after the non-format item in question.

This does basically the same (delete and skip the negated attribute) but without O(N•M) 😅 

A quick step-by-step:

This happens with the current yjs release version (yjs@13.5.27):
```js
  >  initial
   string ID { client: 574277975, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 574277975, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 574277975, clock: 13 } ContentString { str: 'on fire' }
   format ID { client: 574277975, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 574277975, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after findPosition
   string ID { client: 574277975, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 574277975, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 574277975, clock: 13 } ContentString { str: 'on ' }
   string ID { client: 574277975, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 574277975, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 574277975, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after minimizeAttributeChanges
   string ID { client: 574277975, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 574277975, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 574277975, clock: 13 } ContentString { str: 'on ' }
   string ID { client: 574277975, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 574277975, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 574277975, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after insertAttributes
   string ID { client: 574277975, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 574277975, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 574277975, clock: 13 } ContentString { str: 'on ' }
   format ID { client: 574277975, clock: 49 } ContentFormat { key: 'bold', value: null }
   string ID { client: 574277975, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 574277975, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 574277975, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after delete loop and quill check
   string ID { client: 574277975, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 574277975, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 574277975, clock: 13 } ContentString { str: 'on ' }
   format ID { client: 574277975, clock: 49 } ContentFormat { key: 'bold', value: null }
   string ID { client: 574277975, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 574277975, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 574277975, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after insertNegatedAttributes
   string ID { client: 574277975, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 574277975, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 574277975, clock: 13 } ContentString { str: 'on ' }
   format ID { client: 574277975, clock: 49 } ContentFormat { key: 'bold', value: null }
   string ID { client: 574277975, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 574277975, clock: 50 } ContentFormat { key: 'bold', value: true }
   format ID { client: 574277975, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 574277975, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
```

When we just skip the negated attribute (with return) we still get the `{bold: null}` twice:
```js
  >  initial
   string ID { client: 1843338659, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1843338659, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1843338659, clock: 13 } ContentString { str: 'on fire' }
   format ID { client: 1843338659, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1843338659, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after findPosition
   string ID { client: 1843338659, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1843338659, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1843338659, clock: 13 } ContentString { str: 'on ' }
   string ID { client: 1843338659, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 1843338659, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1843338659, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after minimizeAttributeChanges
   string ID { client: 1843338659, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1843338659, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1843338659, clock: 13 } ContentString { str: 'on ' }
   string ID { client: 1843338659, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 1843338659, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1843338659, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after insertAttributes
   string ID { client: 1843338659, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1843338659, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1843338659, clock: 13 } ContentString { str: 'on ' }
   format ID { client: 1843338659, clock: 49 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1843338659, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 1843338659, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1843338659, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after delete loop and quill check
   string ID { client: 1843338659, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1843338659, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1843338659, clock: 13 } ContentString { str: 'on ' }
   format ID { client: 1843338659, clock: 49 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1843338659, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 1843338659, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1843338659, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after insertNegatedAttributes
   string ID { client: 1843338659, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1843338659, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1843338659, clock: 13 } ContentString { str: 'on ' }
   format ID { client: 1843338659, clock: 49 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1843338659, clock: 16 } ContentString { str: 'fire' }
   string ID { client: 1843338659, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
```

With the updated loop (same result my previous solution):
```js
  >  initial
   string ID { client: 1019279783, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1019279783, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1019279783, clock: 13 } ContentString { str: 'on fire' }
   format ID { client: 1019279783, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1019279783, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after findPosition
   string ID { client: 1019279783, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1019279783, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1019279783, clock: 13 } ContentString { str: 'on ' }
   string ID { client: 1019279783, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 1019279783, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1019279783, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after minimizeAttributeChanges
   string ID { client: 1019279783, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1019279783, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1019279783, clock: 13 } ContentString { str: 'on ' }
   string ID { client: 1019279783, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 1019279783, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1019279783, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after insertAttributes
   string ID { client: 1019279783, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1019279783, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1019279783, clock: 13 } ContentString { str: 'on ' }
   format ID { client: 1019279783, clock: 49 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1019279783, clock: 16 } ContentString { str: 'fire' }
   format ID { client: 1019279783, clock: 48 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1019279783, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after delete loop and quill check
   string ID { client: 1019279783, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1019279783, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1019279783, clock: 13 } ContentString { str: 'on ' }
   format ID { client: 1019279783, clock: 49 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1019279783, clock: 16 } ContentString { str: 'fire' }
   string ID { client: 1019279783, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
  >  after insertNegatedAttributes
   string ID { client: 1019279783, clock: 0 } ContentString { str: 'Attack ships ' }
   format ID { client: 1019279783, clock: 47 } ContentFormat { key: 'bold', value: true }
   string ID { client: 1019279783, clock: 13 } ContentString { str: 'on ' }
   format ID { client: 1019279783, clock: 49 } ContentFormat { key: 'bold', value: null }
   string ID { client: 1019279783, clock: 16 } ContentString { str: 'fire' }
   string ID { client: 1019279783, clock: 20 } ContentString { str: ' off the shoulder of Orion.' }
```

dmonad reviewed (changes requested):
Hi @dkuhnert,

Thank you so much for the tests and finding a solution. 

Could you please try to address my comment? I know the source-code is fairly complex and there is little documentation with background knowledge to support. If you can't find a solution for this, I can fix the issue.

Cheers,
Kevin

dkuhnert commented:
Hi @dmonad,

Thanks for your quick response (and the nice library).
I addressed your comment with an new commit. The only gotcha which comes to my head is that now the `currPos` is advanced to the last `ContentFormat` item and the negated attributes are now inserted after it but so far i can't find an issue with that, i think a bit more insight would be good for this type of question 😄.

Thanks for your time!

Cheers,
Dominik

dmonad commented:
Awesome work, thank you so much! 

Gonna get this in the next release tomorrow!
